Add unprefixed Azure request headers support

Korijn opened the pull request:
Adds support for a predefined list of headers which are appropriate for
Azure without the x-ms-meta- prefix.

Resolves issue #1212 and is based on similar pull request #1258.

Korijn committed:
Add unprefixed Azure request headers support

Adds support for a predefined list of headers which are appropriate for
Azure without the x-ms-meta- prefix.

Resolves issue #1212 and is based on similar pull request #1258.

Korijn committed:
Fixed jslint error

Fixed jslint error

Korijn commented:
I tried to sign the CLA but the clahub just barfs "something went wrong" after I fill in the form and submit.

rnicholus commented:
Have you built and tested to make sure this functions exactly as you expect it to?

Korijn committed:
Don't encode parameters that should be as is

For example when setting "Content-MD5" with a value that ends with "==",
it should remain unencoded.

Korijn committed:
Fixed jslint error

Fixed jslint error

Korijn committed:
Fixed another 2 jslint errors

Fixed another 2 jslint errors

Korijn commented:
Admittedly, I was in a rush. I took the time to carefully test it now though. The `Content-MD5` request header can be set using `setParams({ 'Content-MD5', b64_hash }, id)`. Tested in IE11, Firefox 37 and Chrome 42.

![test](https://cloud.githubusercontent.com/assets/1882046/7451036/f2d61a20-f24b-11e4-8c11-135afc969f78.jpg)

Korijn committed:
Improved method docs

Improved method docs

rnicholus commented:
Thanks, I'll merge this into develop to be part of the next scheduled release just as soon as I get some time to write a unit test or two, schedule permitting. If you're able to write a test instead, I'll be able to guarantee that this makes it into the next release.

Korijn committed:
Added unit test

Added unit test for

Korijn commented:
How's this?

Korijn committed:
Corrected jshint errors

Corrected jshint errors

rnicholus commented:
@Korijn Looks good now! Can you take a peek at the [contributor agreement](https://github.com/FineUploader/fine-uploader/blob/master/CONTRIBUTING.md#contributor-agreement) and indicate here if you agree to its terms?

Korijn commented:
Agreed. :+1:

Korijn commented on client/js/azure/util.js:
```diff
@@ -6,11 +6,55 @@ qq.azure.util = qq.azure.util || (function() {
-                var headerName = qq.azure.util.AZURE_PARAM_PREFIX + name;
+                var headerName = qq.azure.util._getPrefixedParamName(name);
 
                 if (qq.isFunction(val)) {
```
I guess this if-else-tree might cause some unexpected behavior... what if the value is a function ánd the parameter name matches an azure parameter? I'll improve upon it next Monday if you agree.

Korijn replied:
I'm a bit short on time for this final adjustment, can we postpone it till after release?

rnicholus replied:
The next scheduled release is probably at least a couple months away. Either way this change can wait, so, not a problem.

On Wed, May 20, 2015 at 3:30 AM, Korijn van Golen
<notifications@github.com> wrote:

>>          getParamsAsHeaders: function(params) {
>>              var headers = {};
>>  
>>              qq.each(params, function(name, val) {
>> -                var headerName = qq.azure.util.AZURE_PARAM_PREFIX + name;
>> +                var headerName = qq.azure.util._getPrefixedParamName(name);
>>  
>>                  if (qq.isFunction(val)) {
> I'm a bit short on time for this final adjustment, can we postpone it till after release?
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/FineUploader/fine-uploader/pull/1404/files#r30680244

Korijn committed:
Improved flow of qq.azure.util.getParamsAsHeaders to handle different
input cases more robustly. Also extended to test to account for this
behaviour.

Korijn committed:
Corrected jscs error.

Korijn commented:
I think the PR is done now.

I could include some more [headers from the Microsoft docs](https://msdn.microsoft.com/en-us/library/azure/dd179451.aspx) but I think you already handle a bunch of them specifically, so I kept the PR restricted to the ones I could test in my setup.

Korijn commented:
I just rebased it on the latest develop (I think) so should be ready to merge if the travis build passes.

Commit:
Corrected for the case of objects being passed in.

rnicholus commented:
Thanks, this looks good. I may end up rolling the hotfix/5.2.3 branch into develop and release this, along with everything currently tagged as 5.2.3 as part of a new 5.3 release (pushing all existing 5.3 stuff to 5.4).
